src/vanilla/utils/proxyMap-indexMap.ts

```diff
@@ -0,0 +1,159 @@
+import { proxy, unstable_getInternalStates } from '../../vanilla.ts'
+const { proxyStateMap } = unstable_getInternalStates()
+const maybeProxify = (x: any) => (typeof x === 'object' ? proxy({ x }).x : x)
+const isProxy = (x: any) => proxyStateMap.has(x)
+
+type InternalProxyObject<K, V> = Map<K, V> & {
+  data: Array<[K, V | undefined]>
+  size: number
+  toJSON: () => Map<K, V>
+}
+
+export function proxyMap<K, V>(entries?: Iterable<[K, V]> | undefined | null) {
+  const data: Array<[K, V]> = []
+  const indexMap = new Map<K, number>()
+  const emptyIndexes: number[] = []
+  let nextIndex = 0
+
+  if (entries !== null && typeof entries !== 'undefined') {
+    if (typeof entries[Symbol.iterator] !== 'function') {
+      throw new TypeError(
+        'proxyMap:\n\tinitial state must be iterable\n\t\ttip: structure should be [[key, value]]',
+      )
+    }
+    for (const [k, v] of entries) {
+      const key = maybeProxify(k)
+      const value = maybeProxify(v)
+      indexMap.set(key, data.length)
+      data.push([key, value])
+    }
+  }
+
+  nextIndex = data.length
+
+  const vObject: InternalProxyObject<K, V> = {
+    data,
+    get size() {
+      return indexMap.size
+    },
+    get(key: K) {
+      const k = maybeProxify(key)
+      if (!indexMap.has(k) && !isProxy(this)) {
+        // eslint-disable-next-line @typescript-eslint/no-unused-expressions
+        this.data.length
+      }
+      if (indexMap.has(k)) {
+        const index = indexMap.get(k)!
+        if (this.data[index] !== undefined) {
+          return this.data[index]![1]
+        }
+      }
+      return undefined
+    },
+    has(key: K) {
+      const k = maybeProxify(key)
+      if (!indexMap.has(k) && !isProxy(this)) {
+        // eslint-disable-next-line @typescript-eslint/no-unused-expressions
+        this.data.length
+      }
+      return indexMap.has(k)
+    },
+    set(key: K, value: V) {
+      if (!isProxy(this)) {
+        if (import.meta.env?.MODE !== 'production') {
+          throw new Error('Cannot perform mutations on a snapshot')
+        } else {
+          return this
+        }
+      }
+      const k = maybeProxify(key)
+      const v = maybeProxify(value)
+      let index = indexMap.get(k)
+      if (index === undefined) {
+        index = emptyIndexes.length ? emptyIndexes.pop()! : nextIndex++
+        indexMap.set(k, index)
+      }
+      const pair = this.data[index]
+      if (pair) {
+        pair[1] = v
+      } else {
+        // Allocate a new array only if necessary
+        this.data[index] = [k, v]
+      }
+      return this
+    },
+    delete(key: K) {
+      if (!isProxy(this)) {
+        if (import.meta.env?.MODE !== 'production') {
+          throw new Error('Cannot perform mutations on a snapshot')
+        } else {
+          return false
+        }
+      }
+      const k = maybeProxify(key)
+      if (indexMap.has(k)) {
+        const index = indexMap.get(k)!
```

Comment:
consistency and micro optimization
```suggestion
      const index = indexMap.get(k)
      if (index !== undefined) {
```